Persistent Menu Navigation in Web Apps

esoergel opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

Adds a pop-out drawer containing a persistent menu.

Collapsed view:
![image](https://github.com/user-attachments/assets/1ef1fe8b-21f5-45f0-abdf-69a8d8f426d1)

Expanded view:
![image](https://github.com/user-attachments/assets/36af7bed-f163-4b96-9aed-4a091891c5d7)


## Technical Summary
[Jira Ticket](https://dimagi.atlassian.net/browse/USH-4711)
[Product Spec](https://docs.google.com/document/d/1MMsRk2aXXTsqPtL5e2Sl9EJdTYGi5mrYfKIls4ct1LU/edit)
[Tech Spec](https://docs.google.com/document/d/12Y4CO-Sl-yTUsdO5VfvDgluRmSXjRTl7FTosQNhyP7M/edit#heading=h.8nqx0ufwj60c)
Branched of of https://github.com/dimagi/commcare-hq/pull/34682

I'd review by commit - commit messages have context for each change.

## Feature Flag
PERSISTENT_MENU_SETTING: Show Persistent Menu option in Web Apps settings

## Safety Assurance

### Safety story

This will be going through QA.  It's a highly visible change, though.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

https://dimagi.atlassian.net/browse/QA-6819

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Wire up rudimentary boilerplate for menu

Please disregard the clumsy B5 positioning stuff

esoergel committed:
Basic toggle functionality

esoergel committed:
Create recursive tree structure for menu cmds

esoergel committed:
Keep track of nested selections, handle click

This uses a weird "trigger" mechanism for handling clicks, since
otherwise parent menu items click events get called too, and we don't
want to call navigate_menu more than once.

esoergel committed:
Navigate on click

esoergel committed:
Bootstrap 5 has a component for this already!

Shout out to Biyeun, whose HTMX demo did something similar, leading me
to this:
https://getbootstrap.com/docs/5.0/components/offcanvas/

esoergel committed:
Display app name and image

esoergel committed:
Style selections better

esoergel committed:
docstrings

esoergel committed:
Restyle toggle button

esoergel committed:
Add images for menu items

esoergel committed:
Make persistent menu images and icons align

esoergel committed:
clear persistent menu when navigating home

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/formplayer_home.html:
```diff
@@ -98,22 +98,24 @@
-      <section id="cloudcare-notifications" class="container notifications-container"></section>
-      <div class="container case-tile-container">
-        <div id="persistent-case-tile" class="print-container"></div>
+      <div id="cloudcare-app-context">
```
Hide whitespace on this commit - I just wrapped an existing block of code

esoergel commented on corehq/apps/hqwebapp/static/cloudcare/scss/formplayer-webapp/menu.scss:
```diff
@@ -12,3 +12,15 @@
+  #persistent-menu-content {
+    background-color: beige;
+  }
+}
```
This was all temporary stuff in helping me build this out - it gets replaced later

esoergel committed:
Remove wrapping div I didn't end up needing

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -161,11 +161,29 @@ hqDefine("cloudcare/js/formplayer/menus/controller", [
             FormplayerFrontend.regions.getRegion('breadcrumb').empty();
         }
+
+        if (menuResponse.persistentMenu) {
```
not at present, but that's the eventual goal.

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/partials/menu/persistent_menu.html:
```diff
@@ -0,0 +1,42 @@
+          class="position-fixed top-50 start-0 rounded-end persistent-menu-toggle">
+    <i class="fa fa-chevron-right"></i>
+  </div>
+  <div id="persistent-menu-container" class="offcanvas offcanvas-start" tabindex="-1" aria-labelledby="persistent-menu-label">
```
I didn't actually test it with a screen reader, but it's supposed to be.  Good to confirm in QA.

esoergel replied:
Did a bunch of fiddling to get the tab navigation stuff right, but it should be good now.

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/partials/menu/persistent_menu.html:
```diff
@@ -0,0 +1,42 @@
+    <% } else { %>
+      <i class="fa fa-arrow-up-right-from-square persistent-menu-image" aria-hidden="true"></i>
+    <% } %>
+    <a class="align-middle" href="#"><%- displayText %></a>
```
I've been browsing the B5 docs and learning a lot!

esoergel committed:
use dashes instead of underscores

To be consistent with
https://github.com/dimagi/formplayer/pull/1594/commits/f774919487a527401b0de9a01771ac46dfc298ce

esoergel committed:
Fix 'selections' array generation

_.union() filters for uniqueness, use concat instead
Also pass in the second arg explicitly - _.each passes in an optional
second arg that is index

esoergel committed:
Dismiss offcanvas explicitly, don't just remove

Removing it doesn't appear to clear whatever it is that freezes the
viewport

esoergel committed:
Parse navigate_menu data to backbone explicitly

This was previously a little piecemeal

esoergel committed:
Let you click the whole row, highlight on hover

esoergel committed:
Get fallback icon from navigationState

esoergel committed:
Make clicking title take you to app main

esoergel committed:
Switch web apps dark blue to medium blue

esoergel committed:
Move navigation bar into cloudcare menu container

esoergel committed:
Move restore-as region into navbar

esoergel committed:
Just use width: 100%

esoergel committed:
Hide offcanvas element when page is redrawn

esoergel committed:
Rename to clarify _which_ menu

esoergel committed:
set isAppPreview once, rather than recalculating

esoergel committed:
Make nomenclature consistent

esoergel committed:
Use different templates for web apps and preview

esoergel committed:
Move hamburger menu into navbar

esoergel committed:
Use bootstrap for breadcrumb divider and mouseover

esoergel committed:
Move breadcrumb colors to app preview and webapps

Since they differ now

esoergel committed:
Darken persistent menu handle on hover

esoergel committed:
Make main logo and breadcrumbs line up

esoergel committed:
Don't load the persistent menu in app preview

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -1489,6 +1495,88 @@ hqDefine("cloudcare/js/formplayer/menus/views", [
+                  icons = {JUMP: 'fa-pencil', NEXT: 'fa-regular fa-folder'};
+            return {
+                imageUri: imageUri ? FormplayerFrontend.getChannel().request('resourceMap', imageUri, appId) : "",
+                iconClass: icons[this.model.get('navigationState')] || 'fa-list-ul',
```
FYI we decided to explicitly make a navigationState for case list. It'll be `ENTITY_SELECT` https://github.com/dimagi/formplayer/pull/1603/commits/0f2f63947b0afb26976fe7fd800b7236103bd10f

esoergel committed:
make persistent menu keyboard navigable

esoergel committed:
Fix breadcrumbs keyboard navigabilitity

I switched to using links in an earlier commit, so I'm updating things
that were earlier done to support clicking the `li` elements

esoergel committed:
Rebuild diffs

esoergel committed:
Bootstrap 5 Migration - Rebuilt diffs

esoergel committed:
Fix keyboard navigation for dropdown menu

This was actually interfering with bootstrap, which handles keyboard
events for dropdowns already.  I don't think using JQuery to add/remove
the css class was functional anymore - guessing it broke with the B5
upgrade

esoergel committed:
lint

AddisonDunn committed:
integrate 'form in progress warning' with persistent menu nav

AddisonDunn committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/persistent-menu

esoergel committed:
reorganize SCSS file to match actual structure

Both of these blocks refer to the same code.  The other selectors for
the first block only apply to live preview, but this is a webapps-only
file.  The btn-group was for the dropdown menu, which has now moved

esoergel committed:
Fix :focus-visible display for breadcrumbs

The overflow controls were on the `li` element, but the `a` element (the
child) was the thing being navigated to.  This meant that bit of the
:focus-visible that overflowed was being hidden - by doing it all on the
same element, things seem to work better.

esoergel committed:
Fix :focus-visible target sizes for navbar

the outlined area was previously bigger than the visible portion of the
link

esoergel committed:
Use explicit entity-select icon

MartinRiese commented on corehq/apps/app_manager/templates/app_manager/profile.xml:
```diff
@@ -29,8 +29,8 @@
 
-    <property key="cc_persistent_menu" value="{{ app.persistent_menu|yesno:"yes,no" }}" force="true"/>
-    <property key="cc_breadcrumbs_enabled" value="{{ app.show_breadcrumbs|yesno:"yes,no" }}" force="true"/>
+    <property key="cc-persistent-menu" value="{{ app.persistent_menu|yesno:"yes,no" }}" force="true"/>
```
Is FP expecting this change?

esoergel replied:
Yep, I made it at Shubham's request

esoergel committed:
Merge pull request #34930 from dimagi/es/webapps-navbar

Reorganize and Restyle Web Apps Navbar

esoergel committed:
rebuild diffs
